head.go

```diff
@@ -601,9 +601,112 @@ func (h *Head) Delete(mint, maxt int64, ms ...labels.Matcher) error {
 	for _, s := range stones {
 		h.tombstones.add(s.ref, s.intervals[0])
 	}
+
+	_, err = h.CleanTombstones()
+	if err != nil {
+		return errors.Wrap(err, "Head: CleanTombstones in Delete")
+	}
+
 	return nil
 }
 
+// CleanTombstones re-writes the chunks with tombstones
+// and deletes chunks if there is no samples after removing tombstones 
+func (h *Head) CleanTombstones() (bool, error) {
+
+	modified := false // flag for entire block
+
+	for sid, dranges := range h.tombstones {
+		if len(dranges) == 0 {
+			continue
+		}
+
+		cleaned := false // flag for current tombstone
+		stripeSeriesIndex := sid%stripeSize
+
+		h.series.locks[stripeSeriesIndex].Lock()
+		s := h.series.series[stripeSeriesIndex] // series with tombstones
+
+		if ms, ok := s[sid]; ok {
+			// Deleting samples for intervals in tombstones
+			for i, chk := range ms.chunks {
+				if !intervalOverlap(dranges[0].Mint, dranges[len(dranges)-1].Maxt, chk.minTime, chk.maxTime) {
+					continue
+				}
+
+				// create new chunk by removing deleted samples
+				newChunk := chunkenc.NewXORChunk()
+				app, err := newChunk.Appender()
+				if err != nil {
+					return modified, err
+				}
+
+				modified = true
+				cleaned = true
+
+				it := &deletedIterator{it: chk.chunk.Iterator(), intervals: dranges}
+				for it.Next() {
+					ts, v := it.At()
+					app.Append(ts, v)
+				}
+
+				if newChunk.NumSamples() > 0 {
+					ms.chunks[i].chunk = newChunk
+				} else { // empty chunk, no required to store
+					ms.chunks = append(ms.chunks[:i], ms.chunks[i+1:]...)
+				}
+			}
+
+			// Getting the last 4 samples in the series
+			// to update the sampleBuf
+			nchks := len(ms.chunks)
+			tofill := 4 // samples left to fill in buffer
+			last4 := []sample{} // last 4 samples
+			for i := nchks-1; i>=0 && tofill>0; i-- {
```

Comment:
Why are you doing this for every chunk? Not just the last chunk?

Reply:
I am iterating from last chunk till I get 4 samples or exhaust the chunks (and not checking all chunks all the time). I had currently taken as `sampleBuf` will have last 4 samples from the series and not just the last chunk. Looking into it now.

Reply:
Seems like updating buffer only the last chunk is enough. I will make the change.